add extradata executor

zayden-nexus opened the pull request:
## Pull Request: ExtraData Executor & SwapBridgeRouter

---

### Title
**feat: Add BridgeExecutor and SwapBridgeRouter for enhanced bridge operations**

---

### Description

This PR introduces two major new contracts and enhances the bridge finalization flow to support external contract calls via `extraData`.

---

### New Contracts

#### 1. BridgeExecutor
A secure executor contract that processes `extraData` during bridge finalization, enabling automated post-bridge actions (DEX swaps, staking, etc.).

**Key Features:**
- **Target Whitelist**: Only pre-approved contracts can be called
- **Method Whitelist** (optional): Per-target function selector restrictions
- **Non-reverting Design**: All failures return `(false, 0)` instead of reverting to prevent DoS attacks
- **Balance Verification**: Always checks token balance before/after external calls
- **Remaining Funds Handling**: Returns unconsumed tokens to the bridge for user refund
- **Gas Protection**: Configurable gas reserve (`_postCallGasReserve`) to ensure post-call cleanup

#### 2. SwapBridgeRouter
A router contract integrating Uniswap V3 swaps with bridge operations in a single transaction.

**Key Features:**
- `swapExactInputSingleAndBridge` / `swapExactInputAndBridge`
- `swapExactOutputSingleAndBridge` / `swapExactOutputAndBridge`
- ETH variants for native token swaps
- Path length validation for V3 encoded paths
- Fee dust refund mechanism

---

### BaseBridge Changes

- **extraData Processing**: Integrated `BridgeExecutor` call during finalization
- **Remaining Funds Transfer**: If executor returns `remaining > 0`, tokens are transferred to the user
- **Low-level Call Safety**: Uses `call` with safe return data parsing to prevent revert propagation
- **Relaxed extraData Length**: Minimum length reduced to 24 bytes (20 address + 4 selector)

---

### Security Improvements

| Issue | Resolution |
|-------|------------|
| Revert propagation DoS | `executeExtraCall` never reverts, returns `(false, 0)` on failure |
| Executor call vulnerability | Low-level `call` with safe bool parsing in BaseBridge |
| Unintended function calls | Optional method ID whitelist per target |
| Unconsumed tokens | Balance verification + remaining funds refund |

---

### New Events

```solidity
// BridgeExecutor
event ExtraCallExecuted(
    uint indexed fromChainID,
    uint indexed index,
    address indexed targetContract,
    bytes4 methodID,
    bool success,
    uint remaining,
    bytes returnData
);
```

---

### Files Changed

| Category | Files |
|----------|-------|
| New Contracts | `BridgeExecutor.sol`, `SwapBridgeRouter.sol` |
| Interfaces | `IBridgeExecutor.sol`, `ISwapBridgeRouter.sol` |
| Modified | `BaseBridge.sol`, `BridgeRegistry.sol` |
| Tests | `BridgeExecutor.t.sol`, `SwapBridgeRouter.t.sol` |
| Dependencies | Uniswap V3 core/periphery submodules |

---

### Testing

```
✅ 135 tests passed, 0 failed
```

New test coverage includes:
- Method whitelist scenarios (allowed/disallowed selectors)
- Remaining funds handling (partial/zero/full consumption)
- Gas exhaustion fallback
- Batch whitelist management
- SwapBridgeRouter quote consistency and swap flows

---

### Breaking Changes

None. The `extraData` feature is opt-in and backward compatible.

zayden-nexus committed:
add extradata executer

zayden-nexus committed:
update

zayden-nexus committed:
update token pause (initiate pause, finalize pause)

luis-nexus commented on src/BaseBridge.sol:
```diff
@@ -151,6 +153,25 @@ contract BaseBridge is
+     * @notice Emitted when the bridge executer is set
+     * @param bridgeExecuter The address of the new bridge executer
+     */
+    event BridgeExecuterSet(address indexed bridgeExecuter);
```
executor?

zayden-nexus replied:
수정 했습니다 ㅎ

joey-nx commented on src/BaseBridge.sol:
```diff
@@ -159,14 +180,17 @@ contract BaseBridge is
     /// @dev Fee management contract
     IBridgeVerifier public bridgeVerifier;
 
+    /// @dev Bridge executer contract for handling extradata operations
+    IBridgeExecuter public bridgeExecuter;
+
     /// @dev dev wallet
     address payable private _dev;
 
     /// @dev Block number when contract was initialized
     uint private _initializedAt;
 
     /// @dev Storage gap for future upgrades
-    uint[47] private __gap;
+    uint[46] private __gap;
```
소스코드 이해를 위해 몇가지 질문 드립니다. 배열 사이즈를 줄이는 이유는 무엇입니까?

joey-nx replied:
@noah-nexus 가 자세히 설명해주셨습니다 ! 감사합니다.

zayden-nexus committed:
fix: rename executer -> executor

noah-nexus commented on src/BridgeExecutor.sol:
```diff
@@ -0,0 +1,190 @@
+     * @param recipient Address to send recovered tokens to
+     */
+    function recoverToken(address token, uint amount, address recipient) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (recipient == address(0)) revert BEInvalidRecipient();
```
validation 하는 김에 token 까지
```suggestion
        if (recipient == address(0) || token == address(0)) revert BEInvalidRecipient();
```

noah-nexus commented on src/BridgeExecutor.sol:
```diff
@@ -0,0 +1,190 @@
+        if (!success) {
+            if (isNative) {
+                (bool sent,) = msg.sender.call{value: value}("");
+                if (!sent) revert BEFailedToReturnNative();
+            } else {
+                toToken.forceApprove(targetContract, 0);
+                // Try to transfer back (works for origin tokens, fails for wrapped)
+                try toToken.transfer(msg.sender, value) {} catch {}
+            }
+        } else if (!isNative) {
+            toToken.forceApprove(targetContract, 0);
+        }
```
중복 코드 제거 

- 실패시 행위와 성공/실패 상관없이 일어나는 행위 분리

```suggestion
        if (!isNative) {
            toToken.forceApprove(targetContract, 0);
        }
        
        if (!success) {
            if (isNative) {
                (bool sent,) = msg.sender.call{value: value}("");
                if (!sent) revert BEFailedToReturnNative();
            } else {
                // Try to transfer back (works for origin tokens, fails for wrapped)
                try toToken.transfer(msg.sender, value) {} catch {}
            }
        }
```

noah-nexus commented on src/BridgeExecutor.sol:
```diff
@@ -0,0 +1,190 @@
+        bool isNative = address(toToken) == Const.NATIVE_TOKEN;
+
+        // Validate balance and approve if needed
+        if (isNative) {
```
방어적 측면에서 value == 0 인 경우에 대한 validation도 추가하면 어떨까요

zayden-nexus replied:
initiate 단계에서 value 0에 대해 제한하고 있습니다.

방어코드로 value 가 0일때 항상 false를 반환하도록 추가했씁니다.

noah-nexus commented on src/BridgeExecutor.sol:
```diff
@@ -0,0 +1,190 @@
+            } else {
+                toToken.forceApprove(targetContract, 0);
+                // Try to transfer back (works for origin tokens, fails for wrapped)
+                try toToken.transfer(msg.sender, value) {} catch {}
```
[사소] : 좀 더 싸게 호출 
```suggestion
        (bool ok, ) = address(toToken).call(
            abi.encodeWithSelector(toToken.transfer.selector, msg.sender, value)
        );
```

noah-nexus commented on src/BaseBridge.sol:
```diff
@@ -666,37 +691,101 @@ contract BaseBridge is
+                // Native token: Executer returned it via call
+                // Origin token: Executer transferred it back
+                // Wrapped token: burn from Executer, then mint to user
+                if (!isOrigin && address(toToken) != Const.NATIVE_TOKEN) {
```
[사소] : 최적화
```suggestion

                bool isERC20 = address(toToken) != Const.NATIVE_TOKEN;

                // For ERC20, transfer/mint to Executer first
                // For Native token, send directly via msg.value in executeExtraCall
                if (isERC20) {
                    status = _transferOrMintToken(toToken, address(bridgeExecuter), value, isOrigin, false);
                    if (status != Const.FinalizeStatus.Success) return status;
                }

                // Call executeExtraCall and check result
                bool success = bridgeExecuter.executeExtraCall{value: isERC20 ? 0 : value}(
                    fromChainID, index, toToken, to, value, extraData
                );

                if (success) {
                    _withdrawToken(fromChainID, address(toToken), value);
                    return Const.FinalizeStatus.Success;
                }

                // Failure: recover tokens based on token type
                emit BridgeExecuterCallFailed(fromChainID, index, toToken, to, value, "");
                // Native token: Executer returned it via call
                // Origin token: Executer transferred it back
                // Wrapped token: burn from Executer, then mint to user
                if (!isOrigin && isERC20) {
```

noah-nexus commented on src/BaseBridge.sol:
```diff
@@ -666,37 +691,101 @@ contract BaseBridge is
+                }
+                status = _transferOrMintToken(toToken, to, value, isOrigin, false);
+                if (status == Const.FinalizeStatus.Success) _withdrawToken(fromChainID, address(toToken), value);
+                return status;
```
756~758과 중복되는 코드 이며 제거해도 동일한 효과를 누릴것이라고 생각되는데 검토 부탁드립니다
```suggestion
                
```

noah-nexus commented on src/BaseBridge.sol:
```diff
@@ -666,37 +691,101 @@ contract BaseBridge is
+        bool isOrigin = _tokenPairs[fromChainID][address(toToken)].isOrigin;
+
+        // Check if extraData is provided and long enough (> 20 bytes for contract address)
+        if (extraData.length > 20 && address(bridgeExecuter) != address(0)) {
```
조건식을 재검토 요청
```suggestion
        if (extraData.length >= 24 && address(bridgeExecuter) != address(0)) {
```

zayden-nexus committed:
Apply review feedback

zayden-nexus committed:
update binding go mod

zayden-nexus committed:
removed check that prevented overwriting crossMintableERC20Code

zayden-nexus committed:
update ExtraCallExecuted event

joey-nx commented on foundry.toml:
```diff
@@ -3,11 +3,11 @@ src = "src"
 libs = ["lib"]
 optimizer = true
-optimizer_runs = 162
+optimizer_runs = 1
```
의도된 변경일까요?

zayden-nexus committed:
add swap bridge router

zayden-nexus committed:
Add Uniswap V3 submodules and update SwapBridgeRouter interface

- Added Uniswap V3 core and periphery submodules to .gitmodules.
- Updated remappings.txt to include paths for Uniswap V3 core and periphery.
- Refactored SwapBridgeRouter to use ISwapRouter and IQuoterV2 interfaces.
- Removed legacy Uniswap V3 interfaces from ISwapBridgeRouter.
- Updated tests to reflect changes in interfaces and ensure correct functionality.
